File: typescript/src/command/command.test.ts
import { describe, expect, it } from "vitest";
import type { Command } from "./command.ts";
import { ConcreteCommand } from "./concrete-command.ts";
import { Invoker } from "./invoker.ts";
import { Receiver } from "./receiver.ts";

describe("Command", () => {
  it("execute and undo act on the receiver", () => {
    const receiver = new Receiver();
    const command = new ConcreteCommand(receiver, "x");
    command.execute();
    expect(receiver.getState()).toBe("x");
    command.undo();
    expect(receiver.getState()).toBe("");
    command.undo(); // nothing left to reverse; harmless
    expect(receiver.getState()).toBe("");
    expect(command.describe()).toBe("ConcreteCommand(x)");
  });

  it("the invoker undoes most recent first", () => {
    const receiver = new Receiver();
    const invoker = new Invoker();
    invoker.execute(new ConcreteCommand(receiver, "a"));
    invoker.execute(new ConcreteCommand(receiver, "b"));
    expect(invoker.undo()?.describe()).toBe("ConcreteCommand(b)");
    expect(receiver.getState()).toBe("a");
    expect(invoker.undo()?.describe()).toBe("ConcreteCommand(a)");
    expect(receiver.getState()).toBe("");
  });

  it("undo with an empty history does nothing", () => {
    expect(new Invoker().undo()).toBeUndefined();
  });

  it("any command works with the invoker", () => {
    const log: string[] = [];
    const custom: Command = {
      execute: () => {
        log.push("do");
      },
      undo: () => {
        log.push("undo");
      },
      describe: () => "custom",
    };
    const invoker = new Invoker();
    invoker.execute(custom);
    invoker.undo();
    expect(log).toEqual(["do", "undo"]);
  });
});

File: java/src/test/java/com/penapereira/patterns/command/CommandTest.java
package com.penapereira.patterns.command;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class CommandTest {

    @Test
    void executeAndUndoActOnTheReceiver() {
        Receiver receiver = new Receiver();
        Command command = new ConcreteCommand(receiver, "x");
        command.execute();
        assertEquals("x", receiver.getState());
        command.undo();
        assertEquals("", receiver.getState());
        command.undo(); // nothing left to reverse; harmless
        assertEquals("", receiver.getState());
        assertEquals("ConcreteCommand(x)", command.describe());
    }

    @Test
    void theInvokerUndoesMostRecentFirst() {
        Receiver receiver = new Receiver();
        Invoker invoker = new Invoker();
        invoker.execute(new ConcreteCommand(receiver, "a"));
        invoker.execute(new ConcreteCommand(receiver, "b"));
        assertEquals("ConcreteCommand(b)", invoker.undo().orElseThrow().describe());
        assertEquals("a", receiver.getState());
        assertEquals("ConcreteCommand(a)", invoker.undo().orElseThrow().describe());
        assertEquals("", receiver.getState());
    }

    @Test
    void undoWithAnEmptyHistoryDoesNothing() {
        assertTrue(new Invoker().undo().isEmpty());
    }

    @Test
    void anyCommandWorksWithTheInvoker() {
        List<String> log = new ArrayList<>();
        Command custom = new Command() {
            @Override
            public void execute() {
                log.add("do");
            }

            @Override
            public void undo() {
                log.add("undo");
            }

            @Override
            public String describe() {
                return "custom";
            }
        };
        Invoker invoker = new Invoker();
        invoker.execute(custom);
        invoker.undo();
        assertEquals(List.of("do", "undo"), log);
    }
}
